packages/zendesk-client/src/use-smooch.ts

```diff
@@ -0,0 +1,121 @@
+import { recordTracksEvent } from '@automattic/calypso-analytics';
+import { useState } from 'react';
+import Smooch from 'smooch';
+import { SMOOCH_INTEGRATION_ID } from './constants';
+import { UserFields } from './types';
+import { useAuthenticateZendeskMessaging } from './use-authenticate-zendesk-messaging';
+import { useUpdateZendeskUserFields } from './use-update-zendesk-user-fields';
+
+export const useSmooch = () => {
+	const [ init, setInit ] = useState( typeof Smooch.getConversations === 'function' );
+	const { data: authData } = useAuthenticateZendeskMessaging( true, 'messenger' );
+	const { isPending: isSubmittingZendeskUserFields, mutateAsync: submitUserFields } =
+		useUpdateZendeskUserFields();
+
+	const initSmooch = ( ref: HTMLDivElement ) => {
+		if ( authData?.jwt && authData?.externalId && ! Smooch.getConversations ) {
+			Smooch.init( {
+				integrationId: SMOOCH_INTEGRATION_ID,
+				embedded: true,
+				externalId: authData?.externalId,
+				jwt: authData?.jwt,
+			} )
+				.then( () => {
+					recordTracksEvent( 'calypso_smooch_messenger_init', {
+						success: true,
+						error: '',
+					} );
+					setInit( true );
+				} )
+				.catch( ( error ) => {
+					recordTracksEvent( 'calypso_smooch_messenger_init', {
+						success: false,
+						error: error.message,
+					} );
+					setInit( false );
+				} );
+			Smooch.render( ref );
+		}
+	};
+
+	const destroy = () => {
+		Smooch.destroy();
+	};
+
+	if ( ! init ) {
+		return {
+			init,
+			destroy,
+			initSmooch,
+			getConversation: async () => undefined,
+			createConversation: async () => undefined,
+			addMessengerListener: () => undefined,
+			addUnreadCountListener: () => undefined,
+			sendMessage: () => undefined,
+		};
+	}
+
+	const getConversation = async ( chatId?: number ): Promise< Conversation | undefined > => {
```

Comment:
It seems all these functions below this line don't depend on any state, but they're still created over and over everytime the hook re-renders. You'd save many cycles if you hoist them up before the hook.

Reply:
Thanks for the feedback.
I've hoisted them up before the hook, except the ones that depend on state.
For the ones that do, I've used `useCallback` to ensure we don't create them over and over.
Let me know if this works.